### DISCUSSION TITLE
WinForms ComboBox with async business rule

### DISCUSSION THREAD

**Discussion by michaelcsikos** [MarimerLLC/csla#4128] [General]
I've just come across some weird behaviour with `ComboBox`, `DropDownStyle = DropDown` and an `async` rule.

If the `AutoCompleteSource = ListItems` and `AutoCompleteMode = SuggestAppend` it's selecting the first match automatically in the `AutoCompleteSource` drop-down, and then the caret moves back to the start, e.g.

type `R`
_[rule completes]_
ComboBox updates to `Red`
type `e`
ComboBox now shows `eRed`

If the `AutoCompleteMode = None` the caret moves to the start as you type a complete match, e.g.
type `R`
type `e`
type `d`
_[caret now positioned before `R`]_

Running the rule sync fixes it, however, in this case I really would prefer to run it `async`, so I dug deeper.

It turns out that if you set `AutoCompleteSource = CustomSource` and add the values to `AutoCompleteCustomSource` it works perfectly.

EDIT: However, you lose the ability to click the drop-down button, of course...

**Comment by rockfordlhotka** [comment]
I hate to say it @michaelcsikos, but you may be one of the few who are still building new(ish) stuff on WinForms.

I'm totally willing to consider a new issue and PR for WinForms capabilities, but personally don't plan to spend time on WinForms.

**Reply by michaelcsikos** [reply]
I'm well aware WinForms is not very popular anymore. Our library of base classes and controls using CSLA + WinForms built up over almost 20 years is hard to beat, though.

I'm not really expecting anyone to dive into solving this one, but maybe someone has run across it before, or can suggest something I could try.

In this particular scenario, I can make the rule sync if the property is bound to a `ComboBox`, and `async` if using a `TextBox`.

**Reply by Chicagoan2016** [reply]
@rockfordlhotka  I am the other one 😂
We have a huge Winforms application and it was much easier to write than another application with WPF UI.
The developer for WPF application didn't use any MVVM pattern but still tried to use the databinding in WPF. It's hell to maintain lol. In his defense, we had looked at Prism and as a small team we didn't have the time to integrate Prism with Csla.
The Winforms application on the other hand is a breeze to maintain. We have discussed this here  #3770

**Reply by Chicagoan2016** [reply]
> In this particular scenario, I can make the rule sync if the property is bound to a `ComboBox`, and `async` if using a `TextBox`.

@michaelcsikos How do you accomplish this, I mean changing rules from ASync to Sycn based on the UI control?
Regards

**Reply by rockfordlhotka** [reply]
I want to be clear.

In no way do I plan to stop having CSLA work with WinForms!! I expect we'll continue to maintain the existing functionality long into the future.

I _think_ what I'm hearing here is that there's some issue that is caused by the fact that the world (and CSLA) have increasingly moved into an async model, while WinForms remains what it has always been - predating the whole async concept entirely.

I am _very_ happy to have contributions to the `Csla.Windows.Forms` package that would help address this, if such a thing is possible.

**Reply by Chicagoan2016** [reply]
@rockfordlhotka, Thank you, Sir. I can't imagine developing an application without using Csla.
Kind Regards

**Reply by michaelcsikos** [reply]
> > In this particular scenario, I can make the rule sync if the property is bound to a `ComboBox`, and `async` if using a `TextBox`.
> 
> @michaelcsikos How do you accomplish this, I mean changing rules from ASync to Sycn based on the UI control? Regards

In this application we've designed a user-defined record type system. One of the options for text fields is to choose the control type: `TextBox`, `ComboBox` with style = `DropDown` (editable tex), or `ComboBox` with style = `DropDownList` (non-editable text).

If the _Unique_ option has been switched on, the app checks if the value has been used in any other project, and if it has displays the number and name in the error provider. In the business rule I already know what control type will be used, so if it's `TextBox` I run the query `async`; if it's `ComboBox` it runs sync. You can still set the rule to be `IsAsync = true` but if you don't `await` anything it runs sync anyway, and the `ComboBox` behaves itself.
